builtin/k8s/status_report_util.go

```diff
@@ -0,0 +1,132 @@
+		resourceHealth.HealthMessage = "container health could not be determined"
+	}
+
+	return &resourceHealth
+}
+
+// Translate a Pod Phase into a Waypoint Health Status
+func podPhaseToHealth(
+	phase corev1.PodPhase,
+) sdk.StatusReport_Health {
+	var healthResult sdk.StatusReport_Health
+
+	switch phase {
+	case corev1.PodPending:
+		healthResult = sdk.StatusReport_ALIVE
+	case corev1.PodRunning:
+		healthResult = sdk.StatusReport_READY
+	case corev1.PodSucceeded:
+		healthResult = sdk.StatusReport_PARTIAL
+	case corev1.PodFailed:
+		healthResult = sdk.StatusReport_DOWN
+	case corev1.PodUnknown:
+		healthResult = sdk.StatusReport_UNKNOWN
+	default:
+		healthResult = sdk.StatusReport_UNKNOWN
+	}
+
+	return healthResult
+}
+
+// Take a list of Pods and build a Waypoint Status Report based on their reported health
+func buildStatusReport(
+	podList *corev1.PodList,
+) sdk.StatusReport {
+	var result sdk.StatusReport
+	result.External = true
+	var resources []*sdk.StatusReport_Resource
+
+	// Build health for every possible pod for overal health report
+	var ready, alive, down, unknown int
+
+	// Report on most recently observed status of a deployments pod
+	// Pod resources and its containers will be in order inside Resources
+	for _, pod := range podList.Items {
+		// Overall Pod Health
+		podStatus := pod.Status
+
+		// Determine overall health report based on all pods
+		switch podStatus.Phase {
+		case corev1.PodPending:
+			alive++
+		case corev1.PodRunning:
+			// Extra checks on the latest condition to ensure pod is reporting ready and running
+			for _, c := range podStatus.Conditions {
+				if c.Status == corev1.ConditionTrue && c.Type == corev1.PodReady {
+					ready++
+					break
+				}
+			}
+
+			alive++
+		case corev1.PodSucceeded:
+			alive++
+		case corev1.PodFailed:
+			down++
+		case corev1.PodUnknown:
+			unknown++
+		default:
+			unknown++
+		}
+
+		podHealth := podPhaseToHealth(podStatus.Phase)
+		resources = append(resources, &sdk.StatusReport_Resource{
+			Health:        podHealth,
+			HealthMessage: podStatus.Message,
+			Name:          pod.ObjectMeta.Name,
+		})
+
+		// Pod containers health
+		for _, containerStatus := range podStatus.ContainerStatuses {
+			resources = append(resources, containerStatusToHealth(containerStatus))
+		}
+	}
+
+	// Overall health status for report
+	if ready == len(podList.Items) {
+		result.Health = sdk.StatusReport_READY
+		result.HealthMessage = "all pods are reporting ready"
+	} else if down == len(podList.Items) {
+		result.Health = sdk.StatusReport_DOWN
+		result.HealthMessage = "all pods are reporting down"
+	} else if unknown == len(podList.Items) {
+		result.Health = sdk.StatusReport_UNKNOWN
+		result.HealthMessage = "status of all pods cannot be determined"
+	} else if alive == len(podList.Items) {
+		result.Health = sdk.StatusReport_ALIVE
+		result.HealthMessage = "all pods are reporting alive"
+	} else {
+		result.Health = sdk.StatusReport_PARTIAL
```

Comment:
Question here: "PARTIAL" implies to me "Partially Ready". With this logic, if 4/5 are down and 1/5 is unknown, we'll get PARTIAL.

Just double-checking that we really want PARTIAL to mean any non-homogeneous set of underlying statuses.

Reply:
Ah, also just found this in the RFC: `PARTIAL = 4; // Some resources in deployment are OK, others are not OK
`